<launch>
    
    <!-- ODOMETRY ARGUMENTS: "type", "nn" and "local_map":
        -Feature type: 0=SURF 1=SIFT 2=ORB 3=FAST/FREAK 4=FAST/BRIEF 5=GFTT/FREAK 6=GFTT/BRIEF 7=BRISK
        -Nearest neighbor strategy : 0=Linear, 1=FLANN_KDTREE, 2=FLANN_LSH, 3=BRUTEFORCE 
           Set to 1 for float descriptor like SIFT/SURF                  
           Set to 3 for binary descriptor like ORB/FREAK/BRIEF/BRISK      
        -Local map size: number of unique features to keep track 
    -->
   <arg name="type" default="6" />
   <arg name="nn" default="3" />
   <arg name="local_map" default="5000" />
   
    <node pkg="camera1394stereo" type="camera1394stereo_node" name="camera1394stereo_node" output="screen" >
	    <param name="video_mode" value="format7_mode3" />
	    <param name="format7_color_coding" value="raw16" />
	    <param name="bayer_pattern" value="bggr" />
	    <param name="bayer_method" value="" />
	    <param name="stereo_method" value="Interlaced" />
	    <param name="camera_info_url_left" value="" />
	    <param name="camera_info_url_right" value="" />
  	</node>
  	
  	<arg name="pi/2" value="1.5707963267948966" />
    <arg name="optical_rotate" value="0 0 0 -$(arg pi/2) 0 -$(arg pi/2)" />
  	<node pkg="tf" type="static_transform_publisher" name="camera_base_link"
        args="$(arg optical_rotate) base_link stereo_camera 100" />  
  	
  	<!-- Run the ROS package stereo_image_proc -->
	<group ns="/stereo_camera" >
		<node pkg="stereo_image_proc" type="stereo_image_proc" name="stereo_image_proc"/>
		
		<!-- Odometry -->
		<node pkg="rtabmap" type="stereo_odometry" name="stereo_odometry" output="screen">
	  	  	<remap from="left/image_rect"       to="left/image_rect_color"/>
	  	  	<remap from="right/image_rect"      to="right/image_rect_color"/>
	  	  	<remap from="left/camera_info"      to="left/camera_info"/>
	  	  	<remap from="right/camera_info"     to="right/camera_info"/>
		  	<remap from="odom" to="odom_stereo"/>
		  	
		  	<remap from="odom_local_map" to="/odom_local_map"/>
		  	<remap from="odom_last_frame" to="/odom_last_frame"/>
	    
	  	  	<param name="frame_id" type="string" value="base_link"/>
	  	  	<param name="odom_frame_id" type="string" value="odom"/>
	  	
	  	  	<param name="min_disparity" type="int" value="0"/>
	  	  	<param name="max_disparity" type="int" value="256"/>
	  	  	<param name="k" type="int" value="10"/>
	  	  	<param name="window_size" type="int" value="5"/>
	  	
	  	  	<param name="Odom/Type" type="string" value="$(arg type)"/>            
  		  	<param name="Odom/NearestNeighbor" type="string" value="$(arg nn)"/>
  		  	<param name="Odom/LocalHistory" type="string" value="$(arg local_map)"/>

  		  	<param name="Odom/MaxDepth" type="string" value="3"/>
	  		<param name="Odom/NNDR" type="string" value="0.6"/>
	  		<param name="GFTT/MaxCorners" type="string" value="1000"/>
	  		<param name="BRIEF/Bytes" type="string" value="16"/>
	  	</node>
		
	</group>
  
  
  <node pkg="rviz" type="rviz" name="rviz" args="-d $(find rtabmap)/launch/config/test_odometry.rviz"/>
  
</launch>